astropy/units/quantity_helper/erfa.py

```diff
@@ -3,15 +3,39 @@
 """Quantity helpers for the ERFA ufuncs."""
 # Tests for these are in coordinates, not in units.
 
-from erfa import ufunc as erfa_ufunc
+from erfa import ufunc as erfa_ufunc, dt_pv, dt_eraLDBODY, dt_eraASTROM
 
 from astropy.units.core import UnitsError, UnitTypeError, dimensionless_unscaled
+from astropy.units.structured import StructuredUnit
 from . import UFUNC_HELPERS
-from .helpers import get_converter, helper_invariant, helper_multiplication
+from .helpers import (get_converter, helper_invariant, helper_multiplication,
+                      helper_twoarg_invariant, _d)
 
 
 erfa_ufuncs = ('s2c', 's2p', 'c2s', 'p2s', 'pm', 'pdp', 'pxp', 'rxp',
-               'gd2gc', 'gc2gd')
+               'cpv', 'p2pv', 'pv2p', 'pv2s', 'pvdpv', 'pvm', 'pvmpv', 'pvppv',
+               'pvstar', 'pvtob', 'pvu', 'pvup', 'pvxpv', 'rxpv', 's2pv', 's2xpv',
+               'starpv', 'sxpv', 'trxpv', 'gd2gc', 'gc2gd', 'ldn', 'aper',
+               'apio', 'atciq', 'atciqn', 'atciqz', 'aticq', 'atioq', 'atoiq')
+
+
+def has_matching_structure(unit, dtype):
+    dtype_fields = dtype.fields
+    if dtype_fields:
+        return (isinstance(unit, StructuredUnit)
+                and len(unit) == len(dtype_fields)
+                and all(has_matching_structure(u, df_v[0])
+                        for (u, df_v) in zip(unit.values(), dtype_fields.values())))
+    else:
+        return not isinstance(unit, StructuredUnit)
+
+
+def check_structured_unit(unit, dtype):
+    if not has_matching_structure(unit, dtype):
+        msg = {dt_pv: 'pv',
```

Comment:
Define the msg outside of this func? It's getting remade every time.

Reply:
Only remade on error, though.